OpenCV build fails on ubuntu with contrib modules

### Please state the information for your system
- OpenCV version: latest master branch for both repos
- Host OS: Linux (Ubuntu 14.04)
- Compiler & CMake: GCC 4.8.5 & CMake 3.4.0-rc1

### In which part of the OpenCV library you got the issue?
Building the repository with extra contrib modules

### Expected behaviour 
The nightly buildbot succeeded in building just fine: http://pullrequest.opencv.org/buildbot/builders/master-contrib_coverage-lin64-debug

### Actual behaviour
It is breaking down at the waldboost code, exactly what I am aiming to build

```
[ 20%] Built target pch_Generate_opencv_perf_videoio
In file included from /home/spu/Documents/github/opencv_compl/build/modules/xobjdetect/precomp.hpp:77:0:
/home/spu/Documents/github/opencv_contrib/modules/xobjdetect/src/waldboost.hpp:53:7: error: redefinition of ‘class cv::xobjdetect::WaldBoost’
 class WaldBoost {
       ^
In file included from /home/spu/Documents/github/opencv_compl/build/modules/xobjdetect/precomp.hpp:48:0:
/usr/local/include/opencv2/xobjdetect.hpp:136:18: error: previous definition of ‘class cv::xobjdetect::WaldBoost’
 class CV_EXPORTS WaldBoost : public Algorithm
                  ^
In file included from /home/spu/Documents/github/opencv_compl/build/modules/xobjdetect/precomp.hpp:78:0:
/home/spu/Documents/github/opencv_contrib/modules/xobjdetect/src/wbdetector.hpp:56:42: error: expected class-name before ‘{’ token
 class WBDetectorImpl : public WBDetector {
                                          ^
/home/spu/Documents/github/opencv_contrib/modules/xobjdetect/src/wbdetector.hpp:71:15: error: cannot declare field ‘cv::xobjdetect::WBDetectorImpl::boost_’ to be of abstract type ‘cv::xobjdetect::WaldBoost’
     WaldBoost boost_;
               ^
In file included from /home/spu/Documents/github/opencv_compl/build/modules/xobjdetect/precomp.hpp:48:0:
/usr/local/include/opencv2/xobjdetect.hpp:136:18: note:   because the following virtual functions are pure within ‘cv::xobjdetect::WaldBoost’:
 class CV_EXPORTS WaldBoost : public Algorithm
                  ^
/usr/local/include/opencv2/xobjdetect.hpp:146:30: note: 	virtual std::vector<int> cv::xobjdetect::WaldBoost::train(cv::Mat&, const cv::Mat&, bool)
     virtual std::vector<int> train(Mat& data,
                              ^
/usr/local/include/opencv2/xobjdetect.hpp:154:19: note: 	virtual float cv::xobjdetect::WaldBoost::predict(const cv::Ptr<cv::xobjdetect::FeatureEvaluator>&) const
     virtual float predict(
```

@alalek could you maybe take a look why it runs fine on the buildbot but not here? I am building with `make -j 24` but also tried single core and it is not a difference, same error.

Here you go!

```
[ 10%] Building CXX object modules/videoio/CMakeFiles/opencv_videoio.dir/src/cap_gstreamer.cpp.o
cd /home/spu/Documents/github/opencv_compl/build/modules/videoio && /usr/bin/c++   -DCVAPI_EXPORTS -DVIDEOIO_EXPORTS -D__OPENCV_BUILD=1 -fsigned-char -W -Wall -Werror=return-type -Werror=non-virtual-dtor -Werror=address -Werror=sequence-point -Wformat -Werror=format-security -Wmissing-declarations -Wundef -Winit-self -Wpointer-arith -Wshadow -Wsign-promo -Wno-narrowing -Wno-delete-non-virtual-dtor -fdiagnostics-show-option -Wno-long-long -pthread -fomit-frame-pointer -msse -msse2 -mno-avx -msse3 -mssse3 -msse4.1 -msse4.2 -ffunction-sections -fvisibility=hidden -fvisibility-inlines-hidden -Wno-deprecated-declarations -O3 -DNDEBUG  -DNDEBUG -fPIC -I/home/spu/Documents/github/opencv_compl/3rdparty/ippicv/unpack/ippicv_lnx/include -I/home/spu/Documents/github/opencv_compl/build -isystem /usr/local/include -isystem /usr/include/gtk-2.0 -isystem /usr/lib/x86_64-linux-gnu/gtk-2.0/include -isystem /usr/include/atk-1.0 -isystem /usr/include/cairo -isystem /usr/include/gdk-pixbuf-2.0 -isystem /usr/include/pango-1.0 -isystem /usr/include/gio-unix-2.0 -isystem /usr/include/freetype2 -isystem /usr/include/glib-2.0 -isystem /usr/lib/x86_64-linux-gnu/glib-2.0/include -isystem /usr/include/pixman-1 -isystem /usr/include/libpng12 -isystem /usr/include/harfbuzz -isystem /usr/include/gstreamer-0.10 -isystem /usr/include/libxml2 -I/home/spu/Documents/github/opencv_compl/modules/videoio/include -I/home/spu/Documents/github/opencv_compl/modules/videoio/src -I/home/spu/Documents/github/opencv_compl/build/modules/videoio -I/home/spu/Documents/github/opencv_compl/modules/cudev/include -I/home/spu/Documents/github/opencv_compl/modules/core/include -I/home/spu/Documents/github/opencv_compl/modules/imgproc/include -I/home/spu/Documents/github/opencv_compl/modules/imgcodecs/include -I/home/spu/Documents/github/opencv_compl/modules/ts/include -I/home/spu/Documents/github/opencv_compl/modules/highgui/include -I/home/spu/Documents/github/opencv_compl/modules/videoio/test -I/home/spu/Documents/github/opencv_compl/modules/videoio/perf     -Winvalid-pch  -o CMakeFiles/opencv_videoio.dir/src/cap_gstreamer.cpp.o -c /home/spu/Documents/github/opencv_compl/modules/videoio/src/cap_gstreamer.cpp
/home/spu/Documents/github/opencv_compl/modules/videoio/src/cap_gstreamer.cpp: In member function ‘virtual bool CvVideoWriter_GStreamer::open(const char*, int, double, CvSize, bool)’:
/home/spu/Documents/github/opencv_compl/modules/videoio/src/cap_gstreamer.cpp:1419:25: error: ‘GST_VIDEO_FORMAT_ENCODED’ was not declared in this scope
         input_pix_fmt = GST_VIDEO_FORMAT_ENCODED;
                         ^
/home/spu/Documents/github/opencv_compl/modules/videoio/src/cap_gstreamer.cpp: In member function ‘virtual bool CvVideoWriter_GStreamer::writeFrame(const IplImage*)’:
/home/spu/Documents/github/opencv_compl/modules/videoio/src/cap_gstreamer.cpp:1590:26: error: ‘GST_VIDEO_FORMAT_ENCODED’ was not declared in this scope
     if (input_pix_fmt == GST_VIDEO_FORMAT_ENCODED) {
                          ^
make[2]: *** [modules/videoio/CMakeFiles/opencv_videoio.dir/src/cap_gstreamer.cpp.o] Error 1
make[2]: Leaving directory `/home/spu/Documents/github/opencv_compl/build'
make[1]: *** [modules/videoio/CMakeFiles/opencv_videoio.dir/all] Error 2
make[1]: Leaving directory `/home/spu/Documents/github/opencv_compl/build'
make: *** [all] Error 2
```

Here you go!

```
[ 51%] Generating precomp.hpp.gch/opencv_xobjdetect_Release.gch
cd /home/spu/Documents/github/opencv_compl/build/modules/xobjdetect && /usr/bin/cmake -E make_directory /home/spu/Documents/github/opencv_compl/build/modules/xobjdetect/precomp.hpp.gch
cd /home/spu/Documents/github/opencv_compl/build/modules/xobjdetect && /usr/bin/c++ -O3 -DNDEBUG -DNDEBUG -fPIC "-D__OPENCV_BUILD=1" "-DCVAPI_EXPORTS" -isystem"/home/spu/Documents/github/opencv_compl/3rdparty/ippicv/unpack/ippicv_lnx/include" -isystem"/home/spu/Documents/github/opencv_compl/build" -isystem"/usr/local/include" -isystem"/home/spu/Documents/github/opencv_contrib/modules/xobjdetect/include" -isystem"/home/spu/Documents/github/opencv_contrib/modules/xobjdetect/src" -isystem"/home/spu/Documents/github/opencv_compl/build/modules/xobjdetect" -I"/home/spu/Documents/github/opencv_compl/modules/cudev/include" -I"/home/spu/Documents/github/opencv_compl/modules/core/include" -I"/home/spu/Documents/github/opencv_compl/modules/imgproc/include" -I"/home/spu/Documents/github/opencv_compl/modules/ml/include" -I"/home/spu/Documents/github/opencv_compl/modules/imgcodecs/include" -I"/home/spu/Documents/github/opencv_compl/modules/videoio/include" -I"/home/spu/Documents/github/opencv_compl/modules/highgui/include" -I"/home/spu/Documents/github/opencv_compl/modules/objdetect/include" -isystem"/home/spu/Documents/github/opencv_compl/3rdparty/ippicv/unpack/ippicv_lnx/include" -isystem"/home/spu/Documents/github/opencv_compl/build" -isystem"/usr/local/include" -isystem"/home/spu/Documents/github/opencv_contrib/modules/xobjdetect/include" -isystem"/home/spu/Documents/github/opencv_contrib/modules/xobjdetect/src" -isystem"/home/spu/Documents/github/opencv_compl/build/modules/xobjdetect" -I"/home/spu/Documents/github/opencv_compl/modules/cudev/include" -I"/home/spu/Documents/github/opencv_compl/modules/core/include" -I"/home/spu/Documents/github/opencv_compl/modules/imgproc/include" -I"/home/spu/Documents/github/opencv_compl/modules/ml/include" -I"/home/spu/Documents/github/opencv_compl/modules/imgcodecs/include" -I"/home/spu/Documents/github/opencv_compl/modules/videoio/include" -I"/home/spu/Documents/github/opencv_compl/modules/highgui/include" -I"/home/spu/Documents/github/opencv_compl/modules/objdetect/include" -fsigned-char -W -Wall -Werror=return-type -Werror=non-virtual-dtor -Werror=address -Werror=sequence-point -Wformat -Werror=format-security -Wmissing-declarations -Wundef -Winit-self -Wpointer-arith -Wshadow -Wsign-promo -Wno-narrowing -Wno-delete-non-virtual-dtor -fdiagnostics-show-option -Wno-long-long -pthread -fomit-frame-pointer -msse -msse2 -mno-avx -msse3 -mssse3 -msse4.1 -msse4.2 -ffunction-sections -fvisibility=hidden -fvisibility-inlines-hidden -DCVAPI_EXPORTS -x c++-header -o /home/spu/Documents/github/opencv_compl/build/modules/xobjdetect/precomp.hpp.gch/opencv_xobjdetect_Release.gch /home/spu/Documents/github/opencv_compl/build/modules/xobjdetect/precomp.hpp
In file included from /home/spu/Documents/github/opencv_compl/build/modules/xobjdetect/precomp.hpp:77:0:
/home/spu/Documents/github/opencv_contrib/modules/xobjdetect/src/waldboost.hpp:53:7: error: redefinition of ‘class cv::xobjdetect::WaldBoost’
 class WaldBoost {
       ^
In file included from /home/spu/Documents/github/opencv_compl/build/modules/xobjdetect/precomp.hpp:48:0:
/usr/local/include/opencv2/xobjdetect.hpp:136:18: error: previous definition of ‘class cv::xobjdetect::WaldBoost’
 class CV_EXPORTS WaldBoost : public Algorithm
                  ^
In file included from /home/spu/Documents/github/opencv_compl/build/modules/xobjdetect/precomp.hpp:78:0:
/home/spu/Documents/github/opencv_contrib/modules/xobjdetect/src/wbdetector.hpp:56:42: error: expected class-name before ‘{’ token
 class WBDetectorImpl : public WBDetector {
                                          ^
/home/spu/Documents/github/opencv_contrib/modules/xobjdetect/src/wbdetector.hpp:71:15: error: cannot declare field ‘cv::xobjdetect::WBDetectorImpl::boost_’ to be of abstract type ‘cv::xobjdetect::WaldBoost’
     WaldBoost boost_;
               ^
In file included from /home/spu/Documents/github/opencv_compl/build/modules/xobjdetect/precomp.hpp:48:0:
/usr/local/include/opencv2/xobjdetect.hpp:136:18: note:   because the following virtual functions are pure within ‘cv::xobjdetect::WaldBoost’:
 class CV_EXPORTS WaldBoost : public Algorithm
                  ^
/usr/local/include/opencv2/xobjdetect.hpp:146:30: note: 	virtual std::vector<int> cv::xobjdetect::WaldBoost::train(cv::Mat&, const cv::Mat&, bool)
     virtual std::vector<int> train(Mat& data,
                              ^
/usr/local/include/opencv2/xobjdetect.hpp:154:19: note: 	virtual float cv::xobjdetect::WaldBoost::predict(const cv::Ptr<cv::xobjdetect::FeatureEvaluator>&) const
     virtual float predict(
                   ^
make[2]: *** [modules/xobjdetect/precomp.hpp.gch/opencv_xobjdetect_Release.gch] Error 1
make[2]: Leaving directory `/home/spu/Documents/github/opencv_compl/build'
make[1]: *** [modules/xobjdetect/CMakeFiles/pch_Generate_opencv_xobjdetect.dir/all] Error 2
make[1]: Leaving directory `/home/spu/Documents/github/opencv_compl/build'
make: *** [all] Error 2
```

--- a/cmake/OpenCVUtils.cmake
+++ b/cmake/OpenCVUtils.cmake
@@ -86,7 +86,9 @@ function(ocv_include_directories)
   set(__add_before "")
   foreach(dir ${ARGN})
     get_filename_component(__abs_dir "${dir}" ABSOLUTE)
-    if("${__abs_dir}" MATCHES "^${OpenCV_SOURCE_DIR}" OR "${__abs_dir}" MATCHES "^${OpenCV_BINARY_DIR}")
+    if("${__abs_dir}" MATCHES "^${OpenCV_SOURCE_DIR}"
+        OR "${__abs_dir}" MATCHES "^${OpenCV_BINARY_DIR}"
+        OR (OPENCV_EXTRA_MODULES_PATH AND "${__abs_dir}" MATCHES "^${OPENCV_EXTRA_MODULES_PATH}"))
       list(APPEND __add_before "${dir}")
     else()
       include_directories(AFTER SYSTEM "${dir}")
@@ -101,7 +103,9 @@ function(ocv_target_include_directories target)
   set(__params "")
   foreach(dir ${ARGN})
     get_filename_component(__abs_dir "${dir}" ABSOLUTE)
-    if("${__abs_dir}" MATCHES "^${OpenCV_SOURCE_DIR}" OR "${__abs_dir}" MATCHES "^${OpenCV_BINARY_DIR}")
+    if("${__abs_dir}" MATCHES "^${OpenCV_SOURCE_DIR}"
+        OR "${__abs_dir}" MATCHES "^${OpenCV_BINARY_DIR}"
+        OR (OPENCV_EXTRA_MODULES_PATH AND "${__abs_dir}" MATCHES "^${OPENCV_EXTRA_MODULES_PATH}"))
       list(APPEND __params "${__abs_dir}")
     else()
       list(APPEND __params "${dir}")

